api/accounts/accounts.go

```diff
@@ -225,27 +225,32 @@ func (a *Accounts) handleCallBatchCode(w http.ResponseWriter, req *http.Request)
 		}
 		return err
 	}
-	results, err := a.batchCall(req.Context(), batchCallData, summary)
+	results, err := a.batchCall(req.Context(), batchCallData, summary, isPending)
 	if err != nil {
 		return err
 	}
 	return utils.WriteJSON(w, results)
 }
 
-func (a *Accounts) batchCall(ctx context.Context, batchCallData *BatchCallData, summary *chain.BlockSummary) (results BatchCallResults, err error) {
+func (a *Accounts) batchCall(ctx context.Context, batchCallData *BatchCallData, summary *chain.BlockSummary, isPending bool) (results BatchCallResults, err error) {
 	txCtx, gas, clauses, err := a.handleBatchCallData(batchCallData)
 	if err != nil {
 		return nil, err
 	}
 	header := summary.Header
 	state := a.stater.NewState(header.StateRoot(), header.Number(), summary.Conflicts, summary.SteadyNum)
 
+	blockNumber := header.Number()
+	if isPending {
+		blockNumber++
+	}
+
 	signer, _ := header.Signer()
 	rt := runtime.New(a.repo.NewChain(header.ParentID()), state,
 		&xenv.BlockContext{
 			Beneficiary: header.Beneficiary(),
 			Signer:      signer,
-			Number:      header.Number(),
+			Number:      blockNumber,
 			Time:        header.Timestamp(),
 			GasLimit:    header.GasLimit(),
 			TotalScore:  header.TotalScore(),
```

Comment:
@otherview I don't think its possible because of this:
```go
state := a.stater.NewState(header.StateRoot(), header.Number(), summary.Conflicts, summary.SteadyNum)
```

I'm getting a not found error when I try building the block manually

Reply:
We might need a bit of refactor around these methods to create a temp block on top of best block instead of modifying the existing best block.

I guess my concern is that it seems like what we're doing is evaluating a call on a duplicated context with a few tweaks. (Best block + change of timestamp + block no).
Won't that make functions that rely on other block properties fragile ?